R/read_csv.R

```diff
@@ -0,0 +1,239 @@
+    if(!param_names_read) {
+      tmp <- gsub("#", "", line, fixed = TRUE)
+      tmp <- gsub("(Default)", "", tmp, fixed = TRUE)
+      key_val <- grep("=", tmp, fixed = TRUE, value = TRUE)
+      if(length(key_val) == 0) {
+        arg_name <- trimws(tmp)
+        if(arg_prefix != "" && !sampling_params_read) {
+          arg_prefix <- paste(arg_prefix, arg_name, sep = "_")
+        } else {
+          arg_prefix <- trimws(tmp)
+        }
+      } else {
+        key_val <- strsplit(key_val, split = "=", fixed = TRUE)
+        key_val <- rapply(key_val, trimws)
+        if(length(key_val) == 2) {
+          numeric_val <- suppressWarnings(as.numeric(key_val[2]))
+          # marks the end of sampling args
+          if(key_val[1] == "id" || key_val[1] == "init") {
+            arg_prefix <- ""
+            sampling_params_read <- TRUE
+          }
+          if(arg_prefix != ""){
+            setting_name <- paste(arg_prefix, key_val[1], sep = "_")
+          } else {
+            setting_name <- key_val[1]
+          }
+          if(!is.na(numeric_val)) {
+            csv_file_info[[setting_name]] <- numeric_val
+          } else {
+            csv_file_info[[setting_name]] <- key_val[2]
+          }
+        }
+      }
+    }
+    if (regexpr("# Diagonal elements of inverse mass matrix:", line, perl = TRUE) > 0) {
+      diagonal_matrix_next <- TRUE
+    } else if(diagonal_matrix_next) {
+      csv_file_info$inverse_mass_matrix_diag <- rapply(strsplit(gsub("# ", "", line), ","), as.numeric)
+      diagonal_matrix_next <- FALSE
+    }
+  }
+  close(con)
+  if(is.null(csv_file_info$inverse_mass_matrix_diag)) {
+    csv_file_info$inverse_mass_matrix_diag <- NULL
+  }
+  if(csv_file_info$method != "sample") {
+    stop("Supplied CSV file was not generated with sampling. Consider using read_optim_csv or read_vb_csv!")
+  }
+  csv_file_info
+}
+
+#' Reads sampling results from the supplied CSV files. Returns a list
+#' containing sampling arguments, the diagonal of the inverse mass
+#' matrix, the post-warmup samples, the sampling parameters and
+#' optionally samples produced during warmup.
+#'
+#' @param output_files Paths to the CSV files to read.
+#'
+#' @return The list of sampling arguments, the diagonal of the inverse mass
+#' matrix, the post-warmup samples, the sampling parameters and warmup samples
+#' if the run was run with save_warmup = 1.
+#'
+read_sample_csv <- function(output_files) {
+  sampling_info <- NULL
+  inverse_mass_matrix_diag <- c()
+  sampling_params_draws <- c()
+  post_warmup_draws_array <- c()
+  warmup_draws_array <- c()
+  for(output_file in output_files) {
+    # read meta data
+    if (is.null(sampling_info)) {
+      sampling_info <- read_sample_info_csv(output_file)
+      inverse_mass_matrix_diag <- sampling_info$inverse_mass_matrix_diag
+    } else {
+      csv_file_info <- read_sample_info_csv(output_file)
+      # check if sampling info matches
+      check_sampling_csv_info_matches(sampling_info,
```

Comment:
The check function throws. Would it be better if it returned the error message in case of an error and we would then throw?

Reply:
Changed to returning error messages.